### PR TITLE
Add toggle to make URL schemes optional

### LINKED ISSUE #7: URL parsing without protocol
I'm looking for something to fetch all URLs from text, and came across this crate. This is brilliant but I'm missing a feature.

I'd like to parse URLs without a protocol definition as well (excluding `https?://`).
So, given this, the following text would yield `example.com` and `example2.com`:
`Some text that references example.com and (example2.com)`

- Yes, this could yield a lot of false links (for text that wasn't intended to be a link), but that's fine for my case.
- I'd prefer to use `linkify`  for this instead of a custom regex, because your crate seems to handle special cases quite well (URLs wrapped within `<>` or `()` and such).

Now, I don't believe this is conform the specs you've linked, but I'm sure this could be useful for others as well. Parsing these links would be disabled by default of course, but could be enabled through a toggle or a new `LinkKind` type.
Are you open for supporting something like this?

To provide some context: I'd like this because I want to parse URLs from Telegram messages. As you can imagine these are written like `example.com` most of the time. Telegram makes these links clickable, but these aren't currently found by `linkify`.

Thanks for the awesome crate by the way, the current interface is perfect!

### PR DESCRIPTION
**Pull request by timvisee**
### Implementation
This PR adds a toggle to make URL schemes such as `https` optional.

Normally only `https://example.com/` would be found in the text below.

```
Link to https://example.com/ and a reference to timvisee.com as well (shameless plug).
```

By explicitly toggling to make schemes optional, the link finder will find `timvisee.com` as well, like this:

```rust
let mut finder = LinkFinder::new();
finder.url_must_have_scheme(false);

finder.links(text).for_each(|link| {
    println!("Found: {:?} (type: {:?})", link.as_str(), link.type());
});

// Found: "https://example.com/" (type: LinkKind::Url)
// Found: "timvisee.com" (type: LinkKind::Url)
```

As seen in the snippet above links with or without a scheme share `LinkKind::Url` to prevent cluttering types for weird (upcoming) variants and edge cases.

### Further notes

Setting the scheme as optional will lead to false positive finds, but it is up to the end user to filter this as discussed in #7.

This implementation will search for both `:` and `.` occurrences if searching with optional schemes. The great thing is that this will find `http://a` without a `.` separator, but when hitting URLs with no scheme a `.` separator is always present for a `host.TLD` combination. This will never find `abc` with no scheme, and no `.` separator. I hope I'm getting across what behavior I'm referring to here.

Because referring to 'https' as a scheme looks more correct, I renamed this feature to 'optional schemes'.

All current tests succeed. With simple manual testing it looks like the library is outputting correct results.

### TODO
This is a work in progress PR, and the following tasks must be completed first before finalizing:

- [x] Add optional scheme toggle
- [x] Correctly find URLs with optional schemes
- [x] Confirm the implementation looks correct (@robinst)
- [x] Update unit tests, test for URLs with no scheme as well
- [x] Ensure URLs aren't found multiple times when searching optional schemes using `.`.
- [x] Rebase on latest `master`
- [x] TLDs should have at least 2 characters

@robinst Feel free to work on the attached branch for any improvements if you desire. Please let me know what you think! I'm quite happy myself with this implementation.

Fixes #7

### CODE REVIEW

File: src/finder.rs [hunk 1]
```diff
@@ -141,6 +142,17 @@ impl LinkFinder {
         self
     }
 
+    /// Define whether URL schemes such as `https` are optional.
+    ///
+    /// By default only URLs having a scheme defined are found.
+    /// By setting this to `true` you make the scheme of URLs optional, to also find URLs like
+    /// `example.org`. For some URLs the used scheme is important, and making the scheme optional
+    /// may lead to finding a lot of false positive URLs.
+    pub fn url_optional_scheme(&mut self, url_optional_scheme: bool) -> &mut LinkFinder {
```
Comment (robinst) [inline src/finder.rs]: Maybe we should name this similar to the existing option `email_domain_must_have_dot`, so `url_must_have_scheme`, which is `true` by default?

**Comment by timvisee** [comment]
> This already looks good.
> 
> From the issue description, I was thinking we'd only want to recognize domains without paths (i.e. `example.org` but not `example.org/foo`). But it actually make sense to recognize full URLs without schemes, as you did here.
> 
> So yeah, please continue :). The most important part is going to be test cases for all the edge cases.

Whoops, didn't mean to be vague. I did want to parse full URLs (just missing the scheme).

Thanks!

File: src/finder.rs (hunk 1 above)
Comment (timvisee) [reply src/finder.rs]: Good point, should have thought about this before. Will change this soon.

**Comment by robinst** [comment]
Hey, AFAICS, only some tests cases are missing. Are you gonna work on those?

**Comment by mre** [comment]
Our link checker, [lychee](https://github.com/lycheeverse/lychee), is using linkify.
I also wanted optional support for skipping this scheme and saw this PR.

Added some tests in my branch that is based on @timvisee's `master`.
You can see the diff [here](https://github.com/robinst/linkify/compare/master...mre:no-protocol-support?expand=1).

Initially I cobbled all non-scheme tests together, but I figured it would make for better 
maintainability to separate the tests similar to the existing tests.
The drawback is that there is some redundancy, but I'd argue that being able to quickly
pinpoint the cause of an error outweighs that drawback. Can refactor as needed.

If I forgot any tests I can add them, otherwise these two things can be ticked off the list:

* [X] Update unit tests, test for URLs with no scheme as well
* [X] Ensure URLs aren't found multiple times when searching optional schemes using ..

Specifically about the second point, links only get detected once in my tests, which is what we were aiming for.
There are currently two tests which are failing for me right now:

```
thread 'international_without_protocol' panicked at 'assertion failed: `(left == right)`
  left: `"üñîçøðé.com/ä"`,
 right: `"|üñîçøðé.com/ä|"`', tests/common/mod.rs:5:5
```

So it looks like this link doesn't get detected without a scheme.

```
thread 'skip_emails_without_protocol' panicked at 'assertion failed: `(left == right)`
  left: `"|foo.bar@example.org|"`,
 right: `"foo.bar@example.org"`', tests/common/mod.rs:5:5
```

this is from #7 and the URL gets detected even though it's a false-positive.

@robinst mentioned what needs to be done in a comment:

> In that case, . would trigger first and we'd have to reject it as a domain, but then still allow @ to detect it as an email. For that to work, we need to make sure things like @ or + are not allowed after a domain.

Help to fix those is greatly appreciated.

Let me know if I should create a separate PR or if you want to merge the changes into `timvisee/master`. Both options are fine with me.

**Comment by timvisee** [comment]
@mre This is awesome, thanks! Sorry that I didn't pick up this work.

I'll merge your changes in my branch, to keep things easy.

**Comment by timvisee** [comment]
I'm seeing compilation errors on CI due to an old Rust version. Should I update the Rust version in this PR as well, or leave it as is? @robinst

**Comment by robinst** [comment]
Wow, nice progress, thanks @mre!

@timvisee I'm having a look at the build now. I'll migrate from Travis to GitHub actions while I'm at it.

**Comment by robinst** [comment]
Done now, please merge or rebase your branch to get a fixed build.

**Comment by timvisee** [comment]
Done! @robinst

**Comment by robinst** [comment]
I had a quick look. Apart from the test cases that need to be fixed, this also currently passes:

```
    assert_linked_without_protocol("example./", "|example./|");
```

I think this should not be recognized. I think we want at least two characters in the last part of the domain. Similar with `foo.example./`, that should not count.

As for the unicode test case, the reason is that the current code to find the "start" is only accepting characters that are valid in the scheme. But the domain part has more valid characters than the scheme. So I think we might need to have a completely separate `find_start` function for the `.` trigger.

**Comment by timvisee** [comment]
Currently don't have a computer available. Just want to say: I believe my branch is open, feel free to commit on it.

**Comment by mre** [comment]
Any update on this?

**Comment by robinst** [comment]
@timvisee Do you have a computer available again? :)

**Comment by mre** [comment]
Ping @timvisee, just in case you have time to work on this. 😅

**Comment by timvisee** [comment]
I'm very sorry. Quite busy at the moment. When I did have time I was distracted by other things.

I'll probably be able to take a look at it in one or two weeks. Feel free to take it over.

File: src/url.rs [hunk 2]
```diff
@@ -29,35 +85,46 @@ impl Scanner for UrlScanner {
...
                     // Check if there's a quote before the scheme,
                     // and stop once we encounter one of those quotes.
                     // https://github.com/robinst/linkify/issues/20
-                    if QUOTES.contains(&c) {
-                        quote = Some(c);
-                    }
-                    break;
+                    quote = Some(c);
+                }
+                c if !has_scheme && !matches!(c, '(' | ')' | '[' | ']' | '{' | '}' | ' ') => {
```
Comment (timvisee) [inline src/url.rs:113]: Somewhat ugly, do you think this is fine?

File: src/url.rs [hunk 3]
```diff
@@ -7,18 +7,74 @@ const QUOTES: &[char] = &['\'', '\"'];
...
+                .chars()
+                .take_while(|c| c.is_ascii_alphabetic())
+                .take(2)
+                .count()
+                < 2;
+            if tld_too_short {
+                return None;
+            }
+
+            // If this looks like an email address, don't scan it as URL
```
Comment (timvisee) [inline src/url.rs:56]: Logic to detect whether an URL (without scheme) is an email address, in which case it won't be recognized as URL.

I'm not too happy about this myself, but don't see a clear way on how to improve this. Do you think this is good enough?

**Comment by timvisee** [comment]
I've finally had the time to go through this! :-)

I believe everything we've talked about is implemented now. Tests are succeeding.

I do have two comments on my implementation, please see the review-comments above.

File: src/url.rs (hunk 2 above)
Comment (mre) [reply src/url.rs:113]: Honestly I think it's fine. If anything, we could move the `matches` part into a separate function (and add an inline attribute). The question is what we'd call the function. 😄 

```rust
#[inline]
/// Checks against characters that can't be the start of a URL
fn is_special_character(c: &char) -> bool {
    matches!(c, '(' | ')' | '[' | ']' | '{' | '}' | ' ')
}
```

Comment (timvisee) [reply src/url.rs:113]: Right! Some sort of negated pattern would have been great here, but there isn't a nice way to express that other than what I've done now, as far as I know.

File: src/url.rs [hunk 4]
```diff
@@ -7,18 +7,74 @@ const QUOTES: &[char] = &['\'', '\"'];
...
+        let (is_slash_slash, separator_len) = if s[separator..].starts_with("://") {
+            (true, "://".len())
+        } else if s[separator..].starts_with('.') {
+            (false, ".".len())
+        } else {
+            return None;
+        };
+        let after_separator = separator + separator_len;
+
+        // When scanning a dot, for URLs without a scheme, do some addadditional checks
```
Comment (mre) [inline src/url.rs:35]: typo: addadditional -> additional

Comment (timvisee) [reply src/url.rs:35]: Fixed. Will wait with pushing until merging to prevent invalidating these review comments.

File: src/url.rs [hunk 5]
```diff
@@ -7,18 +7,74 @@ const QUOTES: &[char] = &['\'', '\"'];
...
+        if !is_slash_slash {
+            // The TLD must be at least 2 characters long
+            let end = &s[after_separator..];
+            let after_last_dot = end
+                .char_indices()
+                .take_while(|(_, c)| *c != '/')
+                .filter(|(_, c)| *c == '.')
+                .last()
+                .map(|(i, _)| i + 1)
+                .unwrap_or(0);
```
Comment (mre) [inline src/url.rs:45]: [rfind](https://doc.rust-lang.org/std/primitive.str.html#method.rfind) could also work:

```rust
let after_last_dot = end.rfind('.').map(|pos| pos+1).unwrap_or(0);
```

This doesn't handle the `/` yet, though. I'd probably strip away that part before.

Comment (timvisee) [reply src/url.rs:45]:
```rust
end.split('/').next().unwrap().rfind('.').map(|pos| pos+1).unwrap_or(0);
```

would likely do the trick.

File: src/url.rs [hunk 6]
```diff
@@ -7,18 +7,74 @@ const QUOTES: &[char] = &['\'', '\"'];
...
+            let tld_too_short = end[after_last_dot..]
+                .chars()
+                .take_while(|c| c.is_ascii_alphabetic())
+                .take(2)
+                .count()
+                < 2;
```
Comment (mre) [inline src/url.rs:51]: The `.take(2)` is not needed since you're checking the `count` against 2 right after.

Comment (timvisee) [reply src/url.rs:51]: It's a micro-optimization, added so it doesn't go through the whole string if it has already found 2 occurrences.

File: src/url.rs (hunk 3 above)
Comment (mre) [reply src/url.rs:56]: Was trying to refactor the code a bit by putting it into a separate function:

```rust
#[inline]
fn is_mail(input: &str) -> bool {
    // Mail address must always contain an `@` sign. Return early if not.
    let at_pos = match input.chars().position(|c| c == '@') {
        Some(at_pos) => at_pos,
        None => return false,
    };
    // TODO: Why is this a special case?
    // Shouldn't colons be allowed in a mail address?
    if input[0..at_pos].contains(':') {
        return false;
    }
    let domain = &input[at_pos + 1..];
    if let Some(dot_pos) = domain.chars().position(|c| c == '.') {
        if self.scan(domain, dot_pos).is_some() {
            return true;
        }
    }
    false
}
```

The idea was to return early to keep the control flow on the left.
Also, maybe we could add some comments to help newcomers understand the logic.
There is still room for improvement I guess. 🤷‍♀️

Comment (mre) [reply src/url.rs:56]: The main advantage would be that we could write unit tests for the mail detection. 😉

File: src/url.rs [hunk 7]
```diff
@@ -29,35 +85,46 @@ impl Scanner for UrlScanner {
 }
 
 impl UrlScanner {
+    // For URL searching starting before the `://` separator, the `has_scheme` parameter should be
+    // true because the URL will have a scheme for sure. If seraching before the `.` separator, it
```
Comment (mre) [inline src/url.rs:89]: seraching -> searching

**Comment by mre** [comment]
Thanks so much for pushing this forward @timvisee. Great work! 🌟

File: src/url.rs (hunk 3 above)
Comment (timvisee) [reply src/url.rs:56]: Good idea to extract it. The `scan` function was getting large.

> // TODO: Why is this a special case?
> // Shouldn't colons be allowed in a mail address?

An URL may contain user credentials in the form of `user:pass@host/`, while an email address cannot. This case might be so uncommon though that we can skip it all together.

File: src/url.rs (hunk 7 above)
Comment (timvisee) [reply src/url.rs:89]: Thanks! I should really fix my Vim spell checker.

File: src/url.rs (hunk 3 above)
Comment (timvisee) [reply src/url.rs:56]: > Shouldn't colons be allowed in an email address?

It seems you're right. According to RFC 532{1,2}

> space and `"(),:;<>@[\]` characters are allowed with restrictions (they are only allowed inside a quoted string, as described in the paragraph below, and in addition, a backslash or double-quote must be preceded by a backslash);

It notes the `:` is only allowed within a quoted string, along with the `@`. This makes handling it quite a bit more complex. I'll just remove the special `:` case, which should make it 'good enough'.

The funny thing is, this library already includes an email address scanner. I'll try to see if I can easily integrate it here first for a simple `is_mail` check without reinventing the wheel.

Comment (timvisee) [reply src/url.rs]: I've added `is_mail` as helper to the email scanner module:

```rust
/// Helper function to check if given string is considered an email address.
#[inline]
pub(crate) fn is_mail(input: &str) -> bool {
    input
        .char_indices()
        .filter(|(_, c)| *c == '@')
        .any(|(i, _)| {
            let scanner = EmailScanner {
                domain_must_have_dot: true,
            };
            scanner.scan(input, i).is_some()
        })
}
```

Now I simply do:

```rust
// If this is an email address, don't scan it as URL
if email::is_mail(&s[after_separator..]) {
    return None;
}
```

This is a lot cleaner.

**Comment by mre** [comment]
@robinst looks like your change request was addressed. Anything else we need to do to get this merged or is this ready to go? 🚀

**Comment by robinst** [comment]
Looks good, merged now! Thanks for your patience. I'll release this shortly.

**Comment by robinst** [comment]
I've added this feature to the demo page now with https://github.com/robinst/linkify/commit/bac4a6412a16fa44e4836a09f5dbdfba9a2da21a: https://robinst.github.io/linkify/

**Comment by robinst** [comment]
Just published version 0.8.0 🎉, thanks for this contribution!: https://github.com/robinst/linkify/blob/main/CHANGELOG.md#080---2021-11-26
